### ISSUE TITLE
RFE: improve systemd-nspawn documentation on how to use it for GUI apps  

### ISSUE THREAD

**Issue by ShalokShalom** [systemd/systemd#12671]
**Is your feature request related to a problem?**

Currently, it's a pain to deliver a graphical application which works distribution independent. 

There are lots and lots of implementations such as Docker and AppImage, while they appeal quite hacky to me, due to their level of integration and focus. 

systemd has the chance to implement that functionality at the base core of the system, where it belongs. And a significant amount of work is already done. 

**Describe the solution you'd like**

A reasonable way to start GUI applications based on a systemd container.

Currently, it takes aeons to direct X11 and Wayland through the layer of abstraction. I think about an abstraction that is similar to [Docker](https://github.com/mviereck/x11docker). 

"Simply" expose the graphical server to the container.

**Comment by refi64** [comment]
Isn't this basically Flatpak, which does proper sandboxing unlike AppImage but is for GUI apps unlike Docker?

**Comment by ShalokShalom** [comment]
Yeah, Flatpak goes into that direction and offers the additional advantage of Pipewire support. The thing with Flatpak is though, that the integration happens not as quite tight to the core of the OS as systemd is.

**Comment by topimiettinen** [comment]
There's already `systemd --user` which can be used for GUI apps too. However, it does not support all features that PID1 can due to kernel limitations. Maybe this will improve some day. D-Bus can be made to use systemd to launch apps.

Firejail is setuid root, so it can do same level of containerization what PID1 can do. There's specific support to GUI sandboxing using Xephyr, Xpra and Xvfb. It can use AppImage files.

**Comment by ShalokShalom** [comment]
Oh, thanks a lot. 

How is the performance with systemd --user?

I read the current implementation to launch graphical apps is very slow.

**Comment by topimiettinen** [comment]
On the contrary, systemd tries to start the apps at the earliest possibility, before some critical components have started (for example things that set up $DISPLAY variable) and then retries failed units rapidly (default 100ms). I've had to slow it down with `RestartSec=1`. This could be solved with more dependencies between user units.

**Comment by ShalokShalom** [comment]
I mean GUI applications in specific. 

The newly announced [SPURV](https://www.collabora.com/news-and-blog/blog/2019/04/01/running-android-next-to-wayland/) also says systemd-nspawn will introduce a lot of overhead. 

This information might be incorrect, I am very interested in your opinion.

**Comment by topimiettinen** [comment]
I have very limited experience with systemd-nspawn or Weston. In principle the overhead depends on the set of features used and the minimum shouldn't be very much. For maximum performance, GUI apps may need direct access to various device drivers and that may need more careful set up or specific bypasses like SPURV. I'd like to see real numbers for performance before agreeing with the statement. I don't see any visible difference with Firejail in video or 3D applications on X11 and systemd-nspawn should be similar.

**Comment by ShalokShalom** [comment]
Even though the specific bypasses like SPURV (which one do you mean?) the perfomance seems to drop quite significantly.

**Comment by topimiettinen** [comment]
I meant that maybe SPURV bypasses something to gain performance that plain systemd-nspawn does not do. Again, we need performance figures to understand which areas need improvement.

I briefly looked at SPURV repositories but I only noticed quick and dirty hacks to for example disable SELinux stuff or seccomp. I doubt that those would be the source of major performance gains, so I probably have missed the critical stuff.

**Comment by topimiettinen** [comment]
Let's ask them. @robertfoss @tomeuv where do you see performance issues with systemd-nspawn?

**Comment by ShalokShalom** [comment]
I mean that they themselves state, that they expect heavy slowdowns due to the fact that they are using systemd-nspawn. Since you have stated you see zero reasons for why this should be the case. I am seriously interested in using one solution for a couple of dozens, maybe hundreds of packages, alongside my already existing distribution specific package manager. So one which provides me with all the goodness of a casual package system, respectively being easily integrated into that one. 

This is at least how I understand [this post](http://0pointer.net/blog/revisiting-how-we-put-together-linux-systems.html) of Lennart.

**Comment by robertfoss** [comment]
Hi! This was unexpected.

Let's clear up the performance comment first.
The performance issues mentioned mostly refer to startup time, which of course is a penalty SPURV has to pay since it boot the entire Android OS in order to run a single App.

As far as I am aware, we were hitting no particular performance issues that were caused by `systemd-nspawn`.

About simplifying a method for doing this, that would be welcome.
Currently we do some ugly stuff:
  - Expose a diverse set of  `/dev/XXX` devices to the guest, including the GPU
  - Expose pulsaudio native unix socket to the guest
  - Expose wayland unix socket to the guest

The full list can be found here: [launch-container.sh](https://gitlab.collabora.com/spurv/device_freedesktop/blob/master/spurv/launch-container.sh#L55)

Some of the above parts are surely more ok than the others, but disregarding the badness of the setup used, it could definitely be simplified. Which would be helpful.

**Comment by ShalokShalom** [comment]
Hi Robert and thanks for your explanation. From a users standpoint: Which tools would you recommend to use systemd right now as an AppImage replacement. So that I can share applications between the same OS, just with different dependencies here and there. So, how to effectively isolate packages that I build now, from those in the mainstream distribution/repositories. 

Thanks a lot

**Comment by robertfoss** [comment]
I'm afraid this isn't a question I'm likely to give a good answer to, I'm
not particularly familiar with AppImage.

**Comment by refi64** [comment]
> It also looks like the app got rewritten several times and still counts around 200.000 lines.
> This suggest that it does reimplement all and everything from scratch 
That's kind of inaccurate, it's C code with a style guide that adds to the line count as well. Flatpak does integrate with cgroups and the like when possible, but at the end of the day, app management and container management aren't necessarily as unified of a concept as they seem. You have to cover distribution, updating, efficient storage, etc, none of which nspawn really touches.

**Comment by refi64** [comment]
> It also looks like the app got rewritten several times and still counts around 200.000 lines.
> This suggest that it does reimplement all and everything from scratch 
That's kind of inaccurate, it's C code with a style guide that adds to the line count as well. Flatpak does integrate with cgroups and the like when possible, but at the end of the day, app management and container management aren't necessarily as unified of a concept as they seem. You have to cover distribution, updating, efficient storage, etc, none of which nspawn really touches.

**Comment by ShalokShalom** [comment]
> I'm afraid this isn't a question I'm likely to give a good answer to, I'm not particularly familiar with AppImage.

I mean simply an image like systemd can create, which I can distribute and reuse on other systems as well. 

> That's kind of inaccurate, it's C code with a style guide that adds to the line count as well.

Interesting. Fine :+1: 

> You have to cover distribution, updating, efficient storage, etc, none of which nspawn really touches. 

This is fine to me, I did not ask for this. :smiley:

**Comment by topimiettinen** [comment]
@robertfoss Here are some suggestions:
- perhaps the script could be simplified by just bind mounting entire /dev
- instead of `sleep 4` you could use `inotifywait`
- script could be divided to first time system initialization which is only needed to be done once (device setup and SELinux policy loading etc) and per application setup and launch. That should speed up launch of a second app or restarting the first one.

**Comment by q2dg** [comment]
I don't know if this resource is rellevant to current discussion (sorry if it's not), but I've found interesting: https://github.com/stken2050/wayland-desktop-container I read it long time ago to try to achieve a graphical systemd-nspawn's container's boot via PXE ("à la LTSP") but I didn't success and gave up.

**Comment by ShalokShalom** [comment]
Thanks a lot :+1:

**Comment by topimiettinen** [comment]
I suppose there's nothing actionable here, so I'm closing this.

**Comment by ShalokShalom** [comment]
Well, documentation, an official statement, screencasts..

**Comment by q2dg** [comment]
Yes...at least some pharagraphs in a man page to guide people...

**Comment by ShalokShalom** [comment]
To give some impression, how to use systemd like Docker.

**Comment by topimiettinen** [comment]
OK, reopening.

**Comment by topimiettinen** [comment]
I also edited the title to hopefully match the intent of RFE better.

**Comment by ShalokShalom** [comment]
Superb :hugs:

**Comment by q2dg** [comment]
There's a nice tuto from https://patrickskiba.com/sysytemd-nspawn/2019/03/21/graphical-applications-in-systemd-nspawn.html but, sadly, my host is running Wayland instead of Xorg, so I don't know how to implement the required changes...
